src/spotify_to_tidal/sync.py

```diff
@@ -172,133 +172,159 @@ def get_tracks_from_spotify_playlist(spotify_session: spotipy.Spotify, spotify_p
     # get all the remaining tracks in parallel
     if results['next']:
         offsets = [ results['limit'] * n for n in range(1, math.ceil(results['total']/results['limit'])) ]
-        extra_results = call_async_with_progress(_get_tracks_from_spotify_playlist, offsets, "",
-                                                 min(len(offsets), 10), spotify_session=spotify_session, playlist_id=spotify_playlist["id"])
+        extra_results = await atqdm.gather( *[asyncio.to_thread(_get_tracks_from_spotify_playlist, offset, spotify_session=spotify_session, playlist_id=spotify_playlist["id"]) for offset in offsets ] )
         for extra_result in extra_results:
             output.extend([r['track'] for r in extra_result['items'] if r['track'] is not None])
     return output
 
-class TidalPlaylistCache:
-    def __init__(self, playlist: tidalapi.Playlist):
-        self._data = playlist.tracks()
+def populate_track_match_cache(spotify_tracks: List[t_spotify.SpotifyTrack], tidal_tracks: List[tidalapi.Track]):
```

Comment:
The following comment I added to the commit message will be helpful in understanding the modifications I made to the algorithm here:

> There are some situations where a given Spotify track can match with
multiple Tidal tracks, or multiple Spotify tracks can match with a
single Tidal track. For example when you have duplicate tracks on the
Spotify playlist which are from single vs. album version so they have
different spotify IDs.

> For example we might have the following two cases for Spotify/Tidal
Case 1:                           Case 2:
A  A' (matches both A and C)      A  A' (matches both A and C)
B  D'                             B  C' (matches both A and C)
C  (matched A')                   C  D'
D                                 D

> The last algorithm failed for case 2 because it would remap C to A'
and so C' would get deleted, which significantly slows the script down.

> The new algorithm handles both cases so long as there was no reordering
occurring. In the case where there *was* reordering it would just mean
that A' and C' could get their order swapped which is a very small
problem and the speed slowdown would have occurred anyway due to
reordering.